File: tablut/search/game.py
"""
Date: 08/11/2020
Author: Carlo Cena

Implementation of method required by tablut game.
"""
import numpy as np
from tablut.utils.bitboards import escapes_bitboard
from tablut.utils.state_utils import build_column
from collections import OrderedDict


class Game:
    def __init__(self, max_time, color, weights):
        self.max_time = max_time
        self.color = color
        self.weights = weights
        self.possible_actions_hor = np.empty(shape=(9, 9), dtype=int)
        self.possible_actions_ver = np.empty(shape=(9, 9), dtype=int)

        # Create possible actions from each position
        self.possible_actions_hor[0][0] = 0b011000000
        self.possible_actions_hor[0][1] = 0b101000000
        self.possible_actions_hor[0][2] = 0b110000000
        self.possible_actions_hor[0][3] = 0b111011111
        self.possible_actions_hor[0][4] = 0b111101111
        self.possible_actions_hor[0][5] = 0b111110111
        self.possible_actions_hor[0][6] = 0b000000011
        self.possible_actions_hor[0][7] = 0b000000101
        self.possible_actions_hor[0][8] = 0b000000110
        self.possible_actions_hor[1][0] = 0b011100000
        self.possible_actions_hor[1][1] = 0b101100000
        self.possible_actions_hor[1][2] = 0b110100000
        self.possible_actions_hor[1][3] = 0b111000000
        self.possible_actions_hor[1][4] = 0b111101111
        self.possible_actions_hor[1][5] = 0b000000111
        self.possible_actions_hor[1][6] = 0b000001011
        self.possible_actions_hor[1][7] = 0b000001101
        self.possible_actions_hor[1][8] = 0b000001110
        self.possible_actions_hor[2][0] = 0b011111111
        self.possible_actions_hor[2][1] = 0b101111111
        self.possible_actions_hor[2][2] = 0b110111111
        self.possible_actions_hor[2][3] = 0b111011111
        self.possible_actions_hor[2][4] = 0b111101111
        self.possible_actions_hor[2][5] = 0b111110111
        self.possible_actions_hor[2][6] = 0b111111011
        self.possible_actions_hor[2][7] = 0b111111101
        self.possible_actions_hor[2][8] = 0b111111110
        self.possible_actions_hor[3][0] = 0b011111110
        self.possible_actions_hor[3][1] = 0b001111110
        self.possible_actions_hor[3][2] = 0b010111110
        self.possible_actions_hor[3][3] = 0b011011110
        self.possible_actions_hor[3][4] = 0b011101110
        self.possible_actions_hor[3][5] = 0b011110110
        self.possible_actions_hor[3][6] = 0b011111010
        self.possible_actions_hor[3][7] = 0b011111100
        self.possible_actions_hor[3][8] = 0b011111110
        self.possible_actions_hor[4][0] = 0b011100000
        self.possible_actions_hor[4][1] = 0b101100000
        self.possible_actions_hor[4][2] = 0b000100000
        self.possible_actions_hor[4][3] = 0b001000000
        self.possible_actions_hor[4][4] = 0b001101100
        self.possible_actions_hor[4][5] = 0b000000100
        self.possible_actions_hor[4][6] = 0b000001000
        self.possible_actions_hor[4][7] = 0b000001101
        self.possible_actions_hor[4][8] = 0b000001110
        self.possible_actions_hor[5] = self.possible_actions_hor[3]
        self.possible_actions_hor[6] = self.possible_actions_hor[2]
        self.possible_actions_hor[7] = self.possible_actions_hor[1]
        self.possible_actions_hor[8] = self.possible_actions_hor[0]

        self.possible_actions_ver = np.transpose(self.possible_actions_hor)

    def produce_actions(self, state):
        """
        Returns a list of possible actions, each action is:
        [k, start_row, start_col, end_row, end_col]
        where
        k == True -> king moves, k==False ->pawn moves
        start_row,start_col -> pieces coordinates
        end_row, end_col -> final coordinates
        """
        action_list = OrderedDict()
        if state.turn == "WHITE":
            r = 0
            while state.king_bitboard[r] == 0:  # Searching king row
                r += 1
            c = 0
            curr_pos_mask = (1 << (8 - c))
            while state.king_bitboard[r] & curr_pos_mask != curr_pos_mask:  # Searching king column
                c += 1
                curr_pos_mask = (1 << (8 - c))
            # First look for actions that lead to escapes
            poss_actions_mask = ~state.white_bitboard[r] & self.possible_actions_hor[r][c]  # Horizontal actions
            poss_actions_mask &= ~state.black_bitboard[r]
            i = 1
            tmp_list = OrderedDict()
            if c != 0:
                new_pos_mask = curr_pos_mask << i
                while i <= c and poss_actions_mask & new_pos_mask != 0:  # Actions to the left, checkers cannot jump
                    tmp_list[(True, r, c, r, c - i)] = None
                    i += 1
                    if i <= c:
                        new_pos_mask = new_pos_mask << 1
                for act in dict(reversed(list(tmp_list.items()))):
                    action_list[act] = None
            i = 1
            tmp_list = OrderedDict()
            if c != 8:
                new_pos_mask = curr_pos_mask >> i  # Actions to the right, checkers cannot jump
                while i <= (8 - c) and poss_actions_mask & new_pos_mask != 0:
                    tmp_list[(True, r, c, r, c + i)] = None
                    i += 1
                    if i <= (8 - c):
                        new_pos_mask = new_pos_mask >> 1
                for act in dict(reversed(list(tmp_list.items()))):
                    action_list[act] = None
            white_column = build_column(state.white_bitboard, curr_pos_mask)  # Building column given position
            black_column = build_column(state.black_bitboard, curr_pos_mask)
            poss_actions_mask = ~white_column & self.possible_actions_ver[r][c]  # Vertical actions
            poss_actions_mask &= ~black_column
            i = 1
            tmp_list = OrderedDict()
            if r != 0:
                new_pos_mask = curr_pos_mask << i
                while i <= r and poss_actions_mask & new_pos_mask != 0:  # Actions up, checkers cannot jump
                    tmp_list[(True, r, c, r - i, c)] = None
                    i += 1
                    if i <= r:
                        new_pos_mask = new_pos_mask << 1
                for act in dict(reversed(list(tmp_list.items()))):
                    action_list[act] = None
            i = 1
            tmp_list = OrderedDict()
            if r != 8:
                new_pos_mask = curr_pos_mask >> i  # Actions down, checkers cannot jump
                while i <= (8 - r) and poss_actions_mask & new_pos_mask != 0:
                    tmp_list[(True, r, c, r + i, c)] = None
                    i += 1
                    if i <= (8 - r):
                        new_pos_mask = new_pos_mask >> 1
                for act in dict(reversed(list(tmp_list.items()))):
                    action_list[act] = None

            for r in range(len(state.white_bitboard)):  # Searching white pawns
                if state.white_bitboard[r] != 0:
                    for c in range(len(state.white_bitboard)):
                        curr_pos_mask = (1 << (8 - c))
                        # If current position is occupied by a white pawn
                        if state.white_bitboard[r] & curr_pos_mask == curr_pos_mask:  # Horizontal moves
                            poss_actions_mask = ~state.white_bitboard[r] & self.possible_actions_hor[r][c]
                            poss_actions_mask &= ~state.king_bitboard[r]
                            poss_actions_mask &= ~state.black_bitboard[r]
                            i = 1
                            if c != 0:
                                new_pos_mask = curr_pos_mask << i  # Actions to the left, checkers cannot jump
                                while i <= c and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r, c - i)] = None
                                    i += 1
                                    if i <= c:
                                        new_pos_mask = new_pos_mask << 1
                            i = 1
                            if c != 8:
                                new_pos_mask = curr_pos_mask >> i  # Actions to the right, checkers cannot jump
                                while i <= (8 - c) and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r, c + i)] = None
                                    i += 1
                                    if i <= (8 - c):
                                        new_pos_mask = new_pos_mask >> 1
                            white_column = build_column(state.white_bitboard,
                                                        curr_pos_mask)  # Building column given position
                            black_column = build_column(state.black_bitboard, curr_pos_mask)
                            king_column = build_column(state.king_bitboard, curr_pos_mask)
                            poss_actions_mask = ~white_column & self.possible_actions_ver[r][c]  # Vertical actions
                            poss_actions_mask &= ~black_column
                            poss_actions_mask &= ~king_column
                            i = 1
                            if r != 0:
                                new_pos_mask = curr_pos_mask << i  # Actions up, checkers cannot jump
                                while i <= r and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r - i, c)] = None
                                    i += 1
                                    if i <= r:
                                        new_pos_mask = new_pos_mask << 1
                            i = 1
                            if r != 8:
                                new_pos_mask = curr_pos_mask >> i  # Actions down, checkers cannot jump
                                while i <= (8 - r) and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r + i, c)] = None
                                    i += 1
                                    if i <= (8 - r):
                                        new_pos_mask = new_pos_mask >> 1

        if state.turn == "BLACK":
            for r in range(len(state.black_bitboard)):  # Searching black pawns
                if state.black_bitboard[r] != 0:
                    for c in range(len(state.black_bitboard)):
                        curr_pos_mask = (1 << (8 - c))
                        # If current position is occupied by a black pawn
                        if state.black_bitboard[r] & curr_pos_mask == curr_pos_mask:  # Horizontal actions
                            poss_actions_mask = ~state.white_bitboard[r] & self.possible_actions_hor[r][c]
                            poss_actions_mask &= ~state.king_bitboard[r]
                            poss_actions_mask &= ~state.black_bitboard[r]
                            i = 1
                            new_pos_mask = curr_pos_mask << i  # Actions to the left, checkers cannot jump
                            while i <= c and poss_actions_mask & new_pos_mask != 0:
                                action_list[(False, r, c, r, c - i)] = None
                                i += 1
                                if i <= c:
                                    new_pos_mask = new_pos_mask << 1
                            i = 1
                            new_pos_mask = curr_pos_mask >> i  # Actions to the right, checkers cannot jump
                            while i <= (8 - c) and poss_actions_mask & new_pos_mask != 0:
                                action_list[(False, r, c, r, c + i)] = None
                                i += 1
                                if i <= (8 - c):
                                    new_pos_mask = new_pos_mask >> 1
                            white_column = build_column(state.white_bitboard,
                                                        curr_pos_mask)  # Building column given position
                            black_column = build_column(state.black_bitboard, curr_pos_mask)
                            king_column = build_column(state.king_bitboard, curr_pos_mask)
                            poss_actions_mask = ~white_column & self.possible_actions_ver[r][c]  # Vertical actions
                            poss_actions_mask &= ~black_column
                            poss_actions_mask &= ~king_column
                            i = 1
                            if r != 0:
                                new_pos_mask = curr_pos_mask << i  # Actions up, checkers cannot jump
                                while i <= r and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r - i, c)] = None
                                    i += 1
                                    if i <= r:
                                        new_pos_mask = new_pos_mask << 1
                            i = 1
                            if r != 8:
                                new_pos_mask = curr_pos_mask >> i  # Actions down, checkers cannot jump
                                while i <= (8 - r) and poss_actions_mask & new_pos_mask != 0:
                                    action_list[(False, r, c, r + i, c)] = None
                                    i += 1
                                    if i <= (8 - r):
                                        new_pos_mask = new_pos_mask >> 1
        return action_list

File: src/thor/Game.java
package thor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import it.unibo.ai.didattica.competition.tablut.domain.State.Turn;

public class Game {
	private int[][] possible_actions_hor = new int[9][9];
	private int[][] possible_actions_ver = new int[9][9];
	private String color;
	
	String getColor() {
		return this.color;
	}

	public Game(int[] weights, String color) {
		super();
		this.color = color;
		this.possible_actions_hor[0][0] = 0b011000000;
		this.possible_actions_hor[0][1] = 0b101000000;
		this.possible_actions_hor[0][2] = 0b110000000;
		this.possible_actions_hor[0][3] = 0b111011111;
		this.possible_actions_hor[0][4] = 0b111101111;
		this.possible_actions_hor[0][5] = 0b111110111;
		this.possible_actions_hor[0][6] = 0b000000011;
		this.possible_actions_hor[0][7] = 0b000000101;
		this.possible_actions_hor[0][8] = 0b000000110;
		this.possible_actions_hor[1][0] = 0b011100000;
		this.possible_actions_hor[1][1] = 0b101100000;
		this.possible_actions_hor[1][2] = 0b110100000;
		this.possible_actions_hor[1][3] = 0b111000000;
		this.possible_actions_hor[1][4] = 0b111101111;
		this.possible_actions_hor[1][5] = 0b000000111;
		this.possible_actions_hor[1][6] = 0b000001011;
		this.possible_actions_hor[1][7] = 0b000001101;
		this.possible_actions_hor[1][8] = 0b000001110;
		this.possible_actions_hor[2][0] = 0b011111111;
		this.possible_actions_hor[2][1] = 0b101111111;
		this.possible_actions_hor[2][2] = 0b110111111;
		this.possible_actions_hor[2][3] = 0b111011111;
		this.possible_actions_hor[2][4] = 0b111101111;
		this.possible_actions_hor[2][5] = 0b111110111;
		this.possible_actions_hor[2][6] = 0b111111011;
		this.possible_actions_hor[2][7] = 0b111111101;
		this.possible_actions_hor[2][8] = 0b111111110;
		this.possible_actions_hor[3][0] = 0b011111110;
		this.possible_actions_hor[3][1] = 0b001111110;
		this.possible_actions_hor[3][2] = 0b010111110;
		this.possible_actions_hor[3][3] = 0b011011110;
		this.possible_actions_hor[3][4] = 0b011101110;
		this.possible_actions_hor[3][5] = 0b011110110;
		this.possible_actions_hor[3][6] = 0b011111010;
		this.possible_actions_hor[3][7] = 0b011111100;
		this.possible_actions_hor[3][8] = 0b011111110;
		this.possible_actions_hor[4][0] = 0b011100000;
		this.possible_actions_hor[4][1] = 0b101100000;
		this.possible_actions_hor[4][2] = 0b000100000;
		this.possible_actions_hor[4][3] = 0b001000000;
		this.possible_actions_hor[4][4] = 0b001101100;
		this.possible_actions_hor[4][5] = 0b000000100;
		this.possible_actions_hor[4][6] = 0b000001000;
		this.possible_actions_hor[4][7] = 0b000001101;
		this.possible_actions_hor[4][8] = 0b000001110;
		this.possible_actions_hor[5] = this.possible_actions_hor[3];
		this.possible_actions_hor[6] = this.possible_actions_hor[2];
		this.possible_actions_hor[7] = this.possible_actions_hor[1];
		this.possible_actions_hor[8] = this.possible_actions_hor[0];
		
		for(int i = 0; i < this.possible_actions_hor.length ; i++)
		{
			for(int j = 0; j < this.possible_actions_hor[0].length; j++)
			{
				this.possible_actions_ver[j][i] = this.possible_actions_hor[i][j];
			}
		}
	}
	
	public List<List<Integer>> produce_actions(BitState state){
		List<List<Integer>> all_actions = new ArrayList<>();
		int new_pos_mask = 0;
		if (state.getTurn() == Turn.WHITE) {
			int r = 0;
		    while (state.getKing_bitboard()[r] == 0) // Searching king row
		    	r += 1;
		    int tmp_r = 8 - r;
		    int c = 0;
		    int tmp_c = 8 - c;
		    int curr_pos_mask = (1 << tmp_c);
		    while ((state.getKing_bitboard()[r] & curr_pos_mask) != curr_pos_mask) {  // Searching king column
		                c += 1;
		                tmp_c = (8 - c);
		                curr_pos_mask = (1 << tmp_c);
		    }
	        // First look for actions that lead to escapes
	        int poss_actions_mask = (~state.getWhite_bitboard()[r] & this.possible_actions_hor[r][c]);  // Horizontal actions
	        poss_actions_mask &= (~state.getBlack_bitboard()[r]);
	        int i = 1;
	        List<List<Integer>> tmp_list = new ArrayList<>();
	        if (c != 0) {
	        	new_pos_mask = curr_pos_mask << i;
	            while (i <= c && (poss_actions_mask & new_pos_mask) != 0) {  // Actions to the left, checkers cannot jump
	            	tmp_list.add(new ArrayList<Integer>(Arrays.asList(1, r, c, r, c-i)));
	                i += 1;
	                if (i <= c)
	                	new_pos_mask = new_pos_mask << 1;
	            }
	            Collections.reverse(tmp_list);
	            all_actions.addAll(tmp_list);
	        }
	        i = 1;
	        tmp_list.clear();
	        if (c != 8) {
	        	new_pos_mask = (curr_pos_mask >> i);  // Actions to the right, checkers cannot jump
	            while (i <= tmp_c && (poss_actions_mask & new_pos_mask) != 0) {
	            	tmp_list.add(new ArrayList<Integer>(Arrays.asList(1, r, c, r, c+i)));
	            	i += 1;
	            	if (i <= tmp_c)
	            		new_pos_mask = new_pos_mask >> 1;
	            }
	            Collections.reverse(tmp_list);
	            all_actions.addAll(tmp_list);
	        }
	        int white_column = Utils.build_column(state.getWhite_bitboard(), curr_pos_mask);  // Building column given position
	        int black_column = Utils.build_column(state.getBlack_bitboard(), curr_pos_mask);
	        curr_pos_mask = (1 << tmp_r);  // Vertical actions
	        poss_actions_mask = (~white_column & this.possible_actions_ver[r][c]);
	        poss_actions_mask &= (~black_column);
	        i = 1;
	        tmp_list.clear();
	        if (r != 0) {
		        new_pos_mask = curr_pos_mask << i;
		        while (i <= r && (poss_actions_mask & new_pos_mask) != 0) {  // Actions up, checkers cannot jump
		        	tmp_list.add(new ArrayList<Integer>(Arrays.asList(1, r, c, r-i, c)));
		            i += 1;
		            if (i <= r)
		            	new_pos_mask = new_pos_mask << 1;
		        }
	        Collections.reverse(tmp_list);
            all_actions.addAll(tmp_list);
	        }
	        i = 1;
	        tmp_list.clear();
	        if (r != 8) {
	        	new_pos_mask = curr_pos_mask >> i;  // Actions down, checkers cannot jump
	            while (i <= tmp_r && (poss_actions_mask & new_pos_mask) != 0) {
	            	tmp_list.add(new ArrayList<Integer>(Arrays.asList(1, r, c, r+i, c)));
	                i += 1;
	                if (i <= tmp_r)
	                	new_pos_mask = new_pos_mask >> 1;
	            }
	            Collections.reverse(tmp_list);
	            all_actions.addAll(tmp_list);
	        }

	        for (r=0; r < state.getWhite_bitboard().length; r++) {  // Searching white pawns
	        	if (state.getWhite_bitboard()[r] != 0) {
	        		tmp_r = 8 - r;
	                for (c=0; c < state.getWhite_bitboard().length; c++) {
	                	curr_pos_mask = (1 << (8 - c));  // Horizontal moves
	                    // If current position is occupied by a white pawn
	                    if ((state.getWhite_bitboard()[r] & curr_pos_mask) == curr_pos_mask) {
	                    	tmp_c = 8 - c;
	                        poss_actions_mask = (~state.getWhite_bitboard()[r] & this.possible_actions_hor[r][c]);
	                        poss_actions_mask &= (~state.getKing_bitboard()[r]);
	                        poss_actions_mask &= (~state.getBlack_bitboard()[r]);
	                        i = 1;
	                        if (c != 0) {
	                        	new_pos_mask = curr_pos_mask << i;  // Actions to the left, checkers cannot jump
	                            while (i <= c && (poss_actions_mask & new_pos_mask) != 0) {
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r, c-i)));
	                                i += 1;
	                                if (i <= c)
	                                	new_pos_mask = new_pos_mask << 1;
	                            }
	                        }
	                        i = 1;
	                        if (c != 8) {
	                        	new_pos_mask = curr_pos_mask >> i;  // Actions to the right, checkers cannot jump
	                            while (i <= tmp_c && (poss_actions_mask & new_pos_mask) != 0) {
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r, c+i)));
	                                i += 1;
	                                if (i <= tmp_c)
	                                	new_pos_mask = new_pos_mask >> 1;
	                            }
	                        }
	                        white_column = Utils.build_column(state.getWhite_bitboard(), curr_pos_mask);  // Building column given position
	                        black_column = Utils.build_column(state.getBlack_bitboard(), curr_pos_mask);
	                        int king_column = Utils.build_column(state.getKing_bitboard(), curr_pos_mask);
	                        curr_pos_mask = (1 << tmp_r);  // Vertical actions
	                        poss_actions_mask = (~white_column & this.possible_actions_ver[r][c]);
	                        poss_actions_mask &= (~black_column);
	                        poss_actions_mask &= (~king_column);
	                        i = 1;
	                        if (r != 0) {
	                        	new_pos_mask = curr_pos_mask << i;  // Actions up, checkers cannot jump
	                            while (i <= r && (poss_actions_mask & new_pos_mask) != 0){
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r-i, c)));
	                                i += 1;
	                                if (i <= r)
	                                	new_pos_mask = new_pos_mask << 1;
	                            }
	                        }
	                        i = 1;
	                        if (r != 8) {
	                        	new_pos_mask = curr_pos_mask >> i;  // Actions down, checkers cannot jump
	                            while (i <= tmp_r && (poss_actions_mask & new_pos_mask) != 0){
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r+i, c)));
	                                i += 1;
	                                if (i <= tmp_r)
	                                	new_pos_mask = new_pos_mask >> 1;
	                            }
	                        }
	                    }
	                }
	        	}
	        }
		}

		else if (state.getTurn() == Turn.BLACK) {
			int r = 0;
		    int c = 0;
		    int tmp_c = 0;
		    int tmp_r = 0;
	        for (r=0; r < state.getBlack_bitboard().length; r++) {  // Searching black pawns
	        	if (state.getBlack_bitboard()[r] != 0) {
	        		tmp_r = 8 - r;
	                for (c=0; c < state.getBlack_bitboard().length; c++) {
	                	int curr_pos_mask = (1 << (8 - c));  // Horizontal moves
	                    // If current position is occupied by a white pawn
	                    if ((state.getBlack_bitboard()[r] & curr_pos_mask) == curr_pos_mask) {
	                    	tmp_c = 8 - c;
	                        int poss_actions_mask = (~state.getWhite_bitboard()[r] & this.possible_actions_hor[r][c]);
	                        poss_actions_mask &= (~state.getKing_bitboard()[r]);
	                        poss_actions_mask &= (~state.getBlack_bitboard()[r]);
	                        int i = 1;
	                        if (c != 0) {
	                        	new_pos_mask = curr_pos_mask << i;  // Actions to the left, checkers cannot jump
	                            while (i <= c && (poss_actions_mask & new_pos_mask) != 0) {
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r, c-i)));
	                                i += 1;
	                                if (i <= c)
	                                	new_pos_mask = new_pos_mask << 1;
	                            }
	                        }
	                        i = 1;
	                        if (c != 8) {
	                        	new_pos_mask = curr_pos_mask >> i;  // Actions to the right, checkers cannot jump
	                            while (i <= tmp_c && (poss_actions_mask & new_pos_mask) != 0) {
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r, c+i)));
	                                i += 1;
	                                if (i <= tmp_c)
	                                	new_pos_mask = new_pos_mask >> 1;
	                            }
	                        }
	                        int white_column = Utils.build_column(state.getWhite_bitboard(), curr_pos_mask);  // Building column given position
	                        int black_column = Utils.build_column(state.getBlack_bitboard(), curr_pos_mask);
	                        int king_column = Utils.build_column(state.getKing_bitboard(), curr_pos_mask);
	                        curr_pos_mask = 1 << tmp_r;  // Vertical actions
	                        poss_actions_mask = (~white_column & this.possible_actions_ver[r][c]);
	                        poss_actions_mask &= (~black_column);
	                        poss_actions_mask &= (~king_column);
	                        i = 1;
	                        if (r != 0) {
	                        	new_pos_mask = curr_pos_mask << i;  // Actions up, checkers cannot jump
	                            while (i <= r && (poss_actions_mask & new_pos_mask) != 0){
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r-i, c)));
	                                i += 1;
	                                if (i <= r)
	                                	new_pos_mask = new_pos_mask << 1;
	                            }
	                        }
	                        i = 1;
	                        if (r != 8) {
	                        	new_pos_mask = curr_pos_mask >> i;  // Actions down, checkers cannot jump
	                            while (i <= tmp_r && (poss_actions_mask & new_pos_mask) != 0){
	                            	all_actions.add(new ArrayList<Integer>(Arrays.asList(0, r, c, r+i, c)));
	                                i += 1;
	                                if (i <= tmp_r)
	                                	new_pos_mask = new_pos_mask >> 1;
	                            }
	                        }
	                    }
	                }
	        	}
	        }
		}
		return all_actions;
	}
	
}
